File: ios/App/App/CartoonifyPlugin.swift
import Foundation
import Capacitor
import UIKit

/**
 * CartoonifyPlugin.swift
 * Capacitor iOS Native Plugin bridging JavaScript UI to Swift Core ML Engine.
 */
@objc(CartoonifyPlugin)
public class CartoonifyPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CartoonifyPlugin"
    public let jsName = "CartoonifyNative"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectFace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "cartoonifyImage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "generateAvatarGrid", returnType: CAPPluginReturnPromise)
    ]

    private let engine = CoreMLCartoonEngine.shared

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve([
            "available": true,
            "platform": "iOS",
            "neuralEngine": true,
            "engine": "Core ML + Apple Neural Engine (ANE)",
            "vision": true
        ])
    }

    @objc func detectFace(_ call: CAPPluginCall) {
        guard let base64String = call.getString("imageBase64"),
              let image = imageFromBase64(base64String) else {
            call.reject("Invalid image base64 data")
            return
        }

        engine.detectFace(in: image) { landmarks in
            guard let lm = landmarks else {
                call.resolve([
                    "detected": false,
                    "confidence": 0.0,
                    "angleDeg": 0.0
                ])
                return
            }

            call.resolve([
                "detected": lm.detected,
                "confidence": lm.confidence,
                "angleDeg": lm.angleDeg,
                "boundingBox": [
                    "x": lm.boundingBox.origin.x,
                    "y": lm.boundingBox.origin.y,
                    "width": lm.boundingBox.size.width,
                    "height": lm.boundingBox.size.height
                ],
                "leftEye": ["x": lm.leftEye.x, "y": lm.leftEye.y],
                "rightEye": ["x": lm.rightEye.x, "y": lm.rightEye.y],
                "noseTip": ["x": lm.noseTip.x, "y": lm.noseTip.y],
                "mouthCenter": ["x": lm.mouthCenter.x, "y": lm.mouthCenter.y],
                "chin": ["x": lm.chin.x, "y": lm.chin.y],
                "forehead": ["x": lm.forehead.x, "y": lm.forehead.y]
            ])
        }
    }

    @objc func cartoonifyImage(_ call: CAPPluginCall) {
        guard let base64String = call.getString("imageBase64"),
              let image = imageFromBase64(base64String) else {
            call.reject("Invalid image data")
            return
        }

        let style = call.getString("style") ?? "anime"
        let warpFactor = call.getFloat("warpFactor") ?? 0.3
        let chinTaper = call.getFloat("chinTaper") ?? 0.25
        let eyeMagnify = call.getFloat("eyeMagnify") ?? 0.2
        let enableFaceAlign = call.getBool("enableFaceAlign") ?? true

        engine.processCartoonify(
            image: image,
            style: style,
            warpFactor: warpFactor,
            chinTaper: chinTaper,
            eyeMagnify: eyeMagnify,
            enableFaceAlign: enableFaceAlign
        ) { [weak self] resultImage in
            guard let self = self, let resultBase64 = self.imageToBase64(resultImage) else {
                call.reject("Failed to encode processed image")
                return
            }

            call.resolve([
                "success": true,
                "style": style,
                "imageBase64": resultBase64,
                "engine": "CoreML (ANE)"
            ])
        }
    }

    @objc func generateAvatarGrid(_ call: CAPPluginCall) {
        guard let base64String = call.getString("imageBase64"),
              let image = imageFromBase64(base64String) else {
            call.reject("Invalid image data")
            return
        }

        let styles = ["anime", "caricature", "pixar3d", "comic_book"]
        let group = DispatchGroup()
        var results: [String: String] = [:]
        let lock = NSLock()

        for style in styles {
            group.enter()
            engine.processCartoonify(
                image: image,
                style: style,
                warpFactor: style == "caricature" ? 0.65 : 0.2,
                chinTaper: style == "caricature" ? 0.45 : 0.2,
                eyeMagnify: 0.25,
                enableFaceAlign: true
            ) { [weak self] resImg in
                if let b64 = self?.imageToBase64(resImg) {
                    lock.lock()
                    results[style] = b64
                    lock.unlock()
                }
                group.leave()
            }
        }

        group.notify(queue: .main) {
            call.resolve([
                "success": true,
                "avatars": results
            ])
        }
    }

    // MARK: - Helpers
    private func imageFromBase64(_ base64: String) -> UIImage? {
        var cleanBase64 = base64
        if let commaIndex = base64.firstIndex(of: ",") {
            cleanBase64 = String(base64[base64.index(after: commaIndex)...])
        }
        guard let data = Data(base64Encoded: cleanBase64, options: .ignoreUnknownCharacters) else { return nil }
        return UIImage(data: data)
    }

    private func imageToBase64(_ image: UIImage) -> String? {
        guard let jpegData = image.jpegData(compressionQuality: 0.90) else { return nil }
        return "data:image/jpeg;base64," + jpegData.base64EncodedString()
    }
}

File: android/app/src/main/java/com/cartoonify/ai/CartoonifyPlugin.kt
package com.cartoonify.ai

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

/**
 * CartoonifyPlugin.kt
 * Capacitor Android Native Plugin bridging JavaScript UI to Kotlin TFLite Engine.
 */
@CapacitorPlugin(name = "CartoonifyNative")
class CartoonifyPlugin : Plugin() {

    private lateinit var engine: TFLiteCartoonEngine
    private val coroutineScope = CoroutineScope(Dispatchers.Main)

    override fun load() {
        super.load()
        engine = TFLiteCartoonEngine(context)
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        val status = engine.getEngineStatus()
        val ret = JSObject().apply {
            put("available", true)
            put("platform", "Android")
            put("gpuActive", status["gpuActive"])
            put("engine", status["engine"])
        }
        call.resolve(ret)
    }

    @PluginMethod
    fun detectFace(call: PluginCall) {
        val base64 = call.getString("imageBase64")
        if (base64 == null) {
            call.reject("Image base64 data is required")
            return
        }

        coroutineScope.launch {
            val bitmap = decodeBase64ToBitmap(base64)
            if (bitmap == null) {
                call.reject("Invalid image bitmap data")
                return@launch
            }

            val landmarks = engine.detectFace(bitmap)
            val ret = JSObject().apply {
                put("detected", landmarks.detected)
                put("confidence", landmarks.confidence)
                put("angleDeg", landmarks.angleDeg)

                val bbox = JSObject().apply {
                    put("x", landmarks.boundingBox.left)
                    put("y", landmarks.boundingBox.top)
                    put("width", landmarks.boundingBox.width())
                    put("height", landmarks.boundingBox.height())
                }
                put("boundingBox", bbox)

                put("leftEye", JSObject().apply {
                    put("x", landmarks.leftEye.x)
                    put("y", landmarks.leftEye.y)
                })
                put("rightEye", JSObject().apply {
                    put("x", landmarks.rightEye.x)
                    put("y", landmarks.rightEye.y)
                })
                put("noseTip", JSObject().apply {
                    put("x", landmarks.noseTip.x)
                    put("y", landmarks.noseTip.y)
                })
                put("mouthCenter", JSObject().apply {
                    put("x", landmarks.mouthCenter.x)
                    put("y", landmarks.mouthCenter.y)
                })
                put("chin", JSObject().apply {
                    put("x", landmarks.chin.x)
                    put("y", landmarks.chin.y)
                })
            }
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun cartoonifyImage(call: PluginCall) {
        val base64 = call.getString("imageBase64")
        if (base64 == null) {
            call.reject("Image data missing")
            return
        }

        val style = call.getString("style") ?: "anime"
        val warpFactor = call.getFloat("warpFactor") ?: 0.3f
        val chinTaper = call.getFloat("chinTaper") ?: 0.25f
        val eyeMagnify = call.getFloat("eyeMagnify") ?: 0.2f
        val enableFaceAlign = call.getBoolean("enableFaceAlign") ?: true

        coroutineScope.launch {
            val bitmap = decodeBase64ToBitmap(base64)
            if (bitmap == null) {
                call.reject("Could not decode image")
                return@launch
            }

            val resultBitmap = engine.processCartoonify(
                bitmap = bitmap,
                style = style,
                warpFactor = warpFactor,
                chinTaper = chinTaper,
                eyeMagnify = eyeMagnify,
                enableFaceAlign = enableFaceAlign
            )

            val outBase64 = encodeBitmapToBase64(resultBitmap)
            val ret = JSObject().apply {
                put("success", true)
                put("style", style)
                put("imageBase64", outBase64)
                put("engine", "TFLite (GPU/NNAPI)")
            }
            call.resolve(ret)
        }
    }

    @PluginMethod
    fun generateAvatarGrid(call: PluginCall) {
        val base64 = call.getString("imageBase64")
        if (base64 == null) {
            call.reject("Image data missing")
            return
        }

        coroutineScope.launch {
            val bitmap = decodeBase64ToBitmap(base64)
            if (bitmap == null) {
                call.reject("Could not decode image")
                return@launch
            }

            val styles = listOf("anime", "caricature", "pixar3d", "comic_book")
            val avatarsObj = JSObject()

            for (st in styles) {
                val styledBitmap = engine.processCartoonify(
                    bitmap = bitmap,
                    style = st,
                    warpFactor = if (st == "caricature") 0.65f else 0.2f,
                    chinTaper = if (st == "caricature") 0.45f else 0.2f,
                    eyeMagnify = 0.25f,
                    enableFaceAlign = true
                )
                avatarsObj.put(st, encodeBitmapToBase64(styledBitmap))
            }

            val ret = JSObject().apply {
                put("success", true)
                put("avatars", avatarsObj)
            }
            call.resolve(ret)
        }
    }

    private fun decodeBase64ToBitmap(base64Str: String): Bitmap? {
        val clean = if (base64Str.contains(",")) base64Str.substringAfter(",") else base64Str
        val bytes = Base64.decode(clean, Base64.DEFAULT)
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }

    private fun encodeBitmapToBase64(bitmap: Bitmap): String {
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream)
        val byteArray = stream.toByteArray()
        return "data:image/jpeg;base64," + Base64.encodeToString(byteArray, Base64.NO_WRAP)
    }

    override fun handleOnDestroy() {
        engine.close()
        super.handleOnDestroy()
    }
}
